### ISSUE TITLE
document message handler match behavior, add unit test

### ISSUE THREAD

**Issue by garlick** [flux-framework/flux-core#233]
It's not entirely clear from the headers what the reactor/dispatch code does when multiple message handlers match an incoming message, for example a wildcard topic glob and a fully specified topic string.

The header should document this carefully at minimum; ideally there should be a man page once interfaces are settled.

Consider adding a unit test for the documented behavior

**Comment by garlick** [comment]
Fixed by pr #244

**Comment by dongahn** [comment]
Cool! Thanks Jim.

**Comment by garlick** [comment]
Reopening to capture discussion with @trws on flux-discuss (edited down)

> [@trws]  So, I was trying some things today to get a feel for how messaging works in flux. Basically, I wrote a little python module that prints every message, see the bottom for the script.  The behavior I got as a result was somewhat unexpected. Messages print, but for some reason it eats part of the shutdown messaging and hangs flux module remove.

> [@garlick] If a message is matched by multiple handlers, last one registered wins.  When I reworked the reactor interfaces so a handler was no longer required (or allowed) to consume its message argument, I was wondering whether it might make sense to call all matching handlers. What do you think?  Seems like this behavior suprised you.

> [@trws] That certainly explains the situation.  Is that behavior true for all message types, including events?  Having topic-based request handled by only one handler makes sense to me, although having a way to tell the reactor to send it to the next watcher from the first  would be exceptionally useful for writing wrappers around modules,  but I would expect events to be handled by all watchers that match the event. 

> [@garlick] True for all message types I believe.  I'll reopen #233 and continue this discussion there so it gets captured.

**Comment by trws** [comment]
@garlick and I discussed this after the meeting, and it seems that part of my confusion was that I happened across an event that is registered in the same context as a module, which is not a normal condition.  The events are delivered to all subscribing contexts, but each context only once, as I understand it, and that seems reasonable.

Basically this was a misunderstanding on my part, but it did make me think that it would be useful to be able to compose services, override a service and add new functionality for example.  Perhaps this would be better served by opening a new ticket?

**Comment by garlick** [comment]
A new ticket for composability sounds fine.

Are we OK then with not tweaking the current dispatcher behavior (as documented in reactor.h):
> If multiple message watchers match a given message, the most recently
> registered will handle it.  Thus it is possible to register handlers for
> "svc.*" then "svc.foo", and the former will match all methods but "foo".

Maybe some kind of nuanced dispatcher logic will emerge from a discussion on composability?

@grondo did you feel the current behavior was wrong?

**Comment by trws** [comment]
All I can think of wanting is the ability to pass the message on to the next consumer if one exists.  Well, that and ensuring that the service control events, for lack of a better term, are harder to accidentally override as I did.
